### ISSUE TITLE
Both fast and full sync stall (due to mis-timing requests? or event processing order?)

### ISSUE THREAD

**Issue by veox** [ethereum/trinity#502]
* trinity Version: trinity-v0.1.0-alpha.20-840-g3b52bdee (current `master` at commit 3b52bdee).
* OS: Arch Linux
* Python Version (python --version): 3.7.3
* Environment (output of `pip freeze`): [in gist](https://gist.github.com/veox/ab0be400c4565e62c1331b10628c0d91#file-pip-freeze-txt)

### What is wrong?

This uses the single-preferred-peer set-up described in #501. `trinity` is in the initial fast-sync stage of getting block headers/bodies and transaction receipts.

Issue highlighted with newlines below:

```
   DEBUG  04-14 15:25:03   FastChainBodySyncer  Got block bodies for 128/128 headers from ETHPeer <Node(0x053d@79.98.29.154)>, from <BlockHeader #1774509 117f6be0>..<BlockHeader #1774636 7766dacb>                                          
   DEBUG  04-14 15:25:04   FastChainBodySyncer  Requesting block bodies for 128 headers from ETHPeer <Node(0x053d@79.98.29.154)>                                                                  
   DEBUG  04-14 15:25:04        SkeletonSyncer  Skeleton header queue is 128/1536 full                                                                                                              
   DEBUG  04-14 15:25:04      HeaderMeatSyncer  Header Skeleton Gaps: active=0 queued=0 max=50                                                                                                                                                
   DEBUG  04-14 15:25:04   FastChainBodySyncer  (in progress, queued, max size) of bodies, receipts: [(128, 317, 2048), (0, 4096, 4096)]. Write capacity? yes                                                                                 
    INFO  04-14 15:25:04   FastChainBodySyncer  blks=0     txs=0      bps=125  tps=428   elapsed=5.0  head=#1766362 7ffe..6ab7  age=1y6m2w                                                                                                    
   DEBUG  04-14 15:25:04   FastChainBodySyncer  Got block bodies for 128/128 headers from ETHPeer <Node(0x053d@79.98.29.154)>, from <BlockHeader #1774637 d89de87f>..<BlockHeader #1774764 3106366a>                                          
   DEBUG  04-14 15:25:04   FastChainBodySyncer  Requesting block bodies for 128 headers from ETHPeer <Node(0x053d@79.98.29.154)>                                                                                                              
   DEBUG  04-14 15:25:04   FastChainBodySyncer  Got block bodies for 128/128 headers from ETHPeer <Node(0x053d@79.98.29.154)>, from <BlockHeader #1774765 85184232>..<BlockHeader #1774892 f5432c88>                                          
   DEBUG  04-14 15:25:04   FastChainBodySyncer  Requesting block bodies for 61 headers from ETHPeer <Node(0x053d@79.98.29.154)>                               
   DEBUG  04-14 15:25:05   FastChainBodySyncer  Got block bodies for 61/61 headers from ETHPeer <Node(0x053d@79.98.29.154)>, from <BlockHeader #1774893 f8cf8278>..<BlockHeader #1774953 721b415b>  
   DEBUG  04-14 15:25:09        SkeletonSyncer  Skeleton header queue is 128/1536 full                                                                                                                                                        
   DEBUG  04-14 15:25:09      HeaderMeatSyncer  Header Skeleton Gaps: active=0 queued=0 max=50                                                                                                                                                
   DEBUG  04-14 15:25:09   FastChainBodySyncer  (in progress, queued, max size) of bodies, receipts: [(0, 0, 2048), (0, 4096, 4096)]. Write capacity? yes
    INFO  04-14 15:25:09   FastChainBodySyncer  blks=0     txs=0      bps=119  tps=407   elapsed=5.0  head=#1766362 7ffe..6ab7  age=1y6m2w                                                          

   DEBUG  04-14 15:25:09   FastChainBodySyncer  Requesting receipts for 245 headers from ETHPeer <Node(0x053d@79.98.29.154)>                                                                                                                  
   DEBUG  04-14 15:25:10   FastChainBodySyncer  Timed out requesting receipts for 245 headers from ETHPeer <Node(0x053d@79.98.29.154)>                                                                                                        
   DEBUG  04-14 15:25:10  ResponseCandidateStream  Got unexpected Receipts payload from ETHPeer <Node(0x053d@79.98.29.154)>      

   DEBUG  04-14 15:25:10   FastChainBodySyncer  Pausing ETHPeer <Node(0x053d@79.98.29.154)> for 15.0s, for sending 0 receipts                                                                       
   DEBUG  04-14 15:25:14        SkeletonSyncer  Skeleton header queue is 128/1536 full                                                                                                                                                        
   DEBUG  04-14 15:25:14      HeaderMeatSyncer  Header Skeleton Gaps: active=0 queued=0 max=50                                                                                                                                                
   DEBUG  04-14 15:25:14   FastChainBodySyncer  (in progress, queued, max size) of bodies, receipts: [(0, 0, 2048), (0, 4096, 4096)]. Write capacity? yes                                                                                     
    INFO  04-14 15:25:14   FastChainBodySyncer  blks=0     txs=0      bps=113  tps=386   elapsed=5.0  head=#1766362 7ffe..6ab7  age=1y6m2w                                                                                                    
   DEBUG  04-14 15:25:19        SkeletonSyncer  Skeleton header queue is 128/1536 full                                                                                                                                                        
   DEBUG  04-14 15:25:19      HeaderMeatSyncer  Header Skeleton Gaps: active=0 queued=0 max=50                                                                                                                                                
   DEBUG  04-14 15:25:19   FastChainBodySyncer  (in progress, queued, max size) of bodies, receipts: [(0, 0, 2048), (0, 4096, 4096)]. Write capacity? yes   
    INFO  04-14 15:25:19   FastChainBodySyncer  blks=0     txs=0      bps=107  tps=367   elapsed=5.0  head=#1766362 7ffe..6ab7  age=1y6m2w                                                          
   DEBUG  04-14 15:25:24        SkeletonSyncer  Skeleton header queue is 128/1536 full                                                                                                                                                        
   DEBUG  04-14 15:25:24      HeaderMeatSyncer  Header Skeleton Gaps: active=0 queued=0 max=50                                                                                                                                                
   DEBUG  04-14 15:25:24   FastChainBodySyncer  (in progress, queued, max size) of bodies, receipts: [(0, 0, 2048), (0, 4096, 4096)]. Write capacity? yes   
    INFO  04-14 15:25:24   FastChainBodySyncer  blks=0     txs=0      bps=102  tps=349   elapsed=5.0  head=#1766362 7ffe..6ab7  age=1y6m2w                                                          

   DEBUG  04-14 15:25:25   FastChainBodySyncer  Requesting receipts for 245 headers from ETHPeer <Node(0x053d@79.98.29.154)>                                                                                                                  
   DEBUG  04-14 15:25:26   FastChainBodySyncer  Timed out requesting receipts for 245 headers from ETHPeer <Node(0x053d@79.98.29.154)>                                                                                                        
   DEBUG  04-14 15:25:26  ResponseCandidateStream  Got unexpected Receipts payload from ETHPeer <Node(0x053d@79.98.29.154)>           

   DEBUG  04-14 15:25:26   FastChainBodySyncer  Pausing ETHPeer <Node(0x053d@79.98.29.154)> for 15.0s, for sending 0 receipts                                                                       
   DEBUG  04-14 15:25:29        SkeletonSyncer  Skeleton header queue is 128/1536 full                                                                                                                                                        
   DEBUG  04-14 15:25:29      HeaderMeatSyncer  Header Skeleton Gaps: active=0 queued=0 max=50                                                                                                                                                
   DEBUG  04-14 15:25:29   FastChainBodySyncer  (in progress, queued, max size) of bodies, receipts: [(0, 0, 2048), (0, 4096, 4096)]. Write capacity? yes                                                                                     
    INFO  04-14 15:25:29   FastChainBodySyncer  blks=0     txs=0      bps=97   tps=331   elapsed=5.0  head=#1766362 7ffe..6ab7  age=1y6m2w                                                                                                    
    INFO  04-14 15:25:32           ETHPeerPool  Connected peers: 1 inbound, 0 outbound                                                                                                                                                        
    INFO  04-14 15:25:32           ETHPeerPool  Peer subscribers: 4, longest queue: ETHChainTipMonitor(0)                                                                                                                                     
   DEBUG  04-14 15:25:32           ETHPeerPool  == Peer details ==                                                                        
   DEBUG  04-14 15:25:32           ETHPeerPool  ETHPeer <Node(0x053d@79.98.29.154)>: uptime=0:00:05:43, received_msgs=1661, most_received=Transactions (cmd_id=18)(556)                             
   DEBUG  04-14 15:25:32           ETHPeerPool  client_version_string='Geth/v1.8.26-stable-cdae1c59/linux-amd64/go1.10.4'                                                                                                                     
   DEBUG  04-14 15:25:32           ETHPeerPool      BlockBodies: msgs=522  items=329907  rtt=0.16/0.37/0.09  ips=6060.98429  timeouts=7  quality=98                                                                                           
   DEBUG  04-14 15:25:32           ETHPeerPool      BlockHeaders: msgs=309  items=58366  rtt=0.05/0.14/0.02  ips=4044.76922  timeouts=5  quality=99
   DEBUG  04-14 15:25:32           ETHPeerPool      NodeData: None                                                                                                                                  
   DEBUG  04-14 15:25:32           ETHPeerPool      Receipts: msgs=223  items=165505  rtt=0.12/0.35/0.08  ips=3999.75970  timeouts=9  quality=63                                                                                              
   DEBUG  04-14 15:25:32           ETHPeerPool  == End peer details ==
```

Note that it's always the same-size batch. (On one occasion, I've seen batch size increase, as there were more block bodies received in the interim. `trinity` then requests even more transaction receipts, and that keeps timing out, too.)

Note also that this _might_ be remote-specific. (Haven't tested with a `parity` node yet.)

### How can it be fixed

Not sure.

* Split batch/bundle into smaller ones on timeout happening?
* Increase timeout duration on timeout happening?

**Comment by veox** [comment]
The receipt-timeout pattern keeps repeating, and (as expected) degrades that peer's quality:

```
   DEBUG  04-14 15:31:32           ETHPeerPool  == Peer details ==                            
   DEBUG  04-14 15:31:32           ETHPeerPool  ETHPeer <Node(0x053d@79.98.29.154)>: uptime=0:00:11:43, received_msgs=2413, most_received=Transactions (cmd_id=18)(1236)                                                                      
   DEBUG  04-14 15:31:32           ETHPeerPool  client_version_string='Geth/v1.8.26-stable-cdae1c59/linux-amd64/go1.10.4'                                                                                                                     
   DEBUG  04-14 15:31:32           ETHPeerPool      BlockBodies: msgs=522  items=329907  rtt=0.16/0.37/0.09  ips=6060.98429  timeouts=7  quality=98
   DEBUG  04-14 15:31:32           ETHPeerPool      BlockHeaders: msgs=309  items=58366  rtt=0.05/0.14/0.02  ips=4044.76922  timeouts=5  quality=99
   DEBUG  04-14 15:31:32           ETHPeerPool      NodeData: None                                                                                                                                                                            
   DEBUG  04-14 15:31:32           ETHPeerPool      Receipts: msgs=245  items=165505  rtt=0.12/0.35/0.08  ips=1294.05644  timeouts=31  quality=20                                                                                             
   DEBUG  04-14 15:31:32           ETHPeerPool  == End peer details ==
```

**Comment by veox** [comment]
Another run shows this happening with block body requests instead of tx receipts. There's also some funky fiddling with `ResponseCandidateStream` on peer disconnect/re-connect.

Doing another run with separate `trinity` log, and restarting both nodes at the same time; will link when issue encountered.

**Comment by veox** [comment]
A later experiment with modified `datefmt` (return milliseconds) shows this might be due to event ordering or something:

```
   DEBUG  04-14 19:57:05.547   FastChainBodySyncer  Requesting block bodies for 120 headers from ETHPeer <Node(0x053d@79.98.29.154)>
   DEBUG  04-14 19:57:06.048   FastChainBodySyncer  Timed out requesting block bodies for 120 headers from ETHPeer <Node(0x053d@79.98.29.154)>
   DEBUG  04-14 19:57:06.049  ResponseCandidateStream  Got unexpected BlockBodies payload from ETHPeer <Node(0x053d@79.98.29.154)>
--
   DEBUG  04-14 19:57:06.054   FastChainBodySyncer  Requesting block bodies for 128 headers from ETHPeer <Node(0x053d@79.98.29.154)>
   DEBUG  04-14 19:57:06.548   FastChainBodySyncer  Timed out requesting block bodies for 128 headers from ETHPeer <Node(0x053d@79.98.29.154)>
   DEBUG  04-14 19:57:06.550  ResponseCandidateStream  Got unexpected BlockBodies payload from ETHPeer <Node(0x053d@79.98.29.154)>
--
   DEBUG  04-14 19:57:21.560   FastChainBodySyncer  Requesting block bodies for 128 headers from ETHPeer <Node(0x053d@79.98.29.154)>
   DEBUG  04-14 19:57:22.027   FastChainBodySyncer  Timed out requesting block bodies for 128 headers from ETHPeer <Node(0x053d@79.98.29.154)>
   DEBUG  04-14 19:57:22.029  ResponseCandidateStream  Got unexpected BlockBodies payload from ETHPeer <Node(0x053d@79.98.29.154)>
--
   DEBUG  04-14 19:57:39.080   FastChainBodySyncer  Requesting block bodies for 124 headers from ETHPeer <Node(0x053d@79.98.29.154)>
   DEBUG  04-14 19:57:39.560   FastChainBodySyncer  Timed out requesting block bodies for 124 headers from ETHPeer <Node(0x053d@79.98.29.154)>
   DEBUG  04-14 19:57:39.563  ResponseCandidateStream  Got unexpected BlockBodies payload from ETHPeer <Node(0x053d@79.98.29.154)>
--
   DEBUG  04-14 19:58:08.715   FastChainBodySyncer  Requesting block bodies for 126 headers from ETHPeer <Node(0x053d@79.98.29.154)>
   DEBUG  04-14 19:58:09.512   FastChainBodySyncer  Timed out requesting block bodies for 126 headers from ETHPeer <Node(0x053d@79.98.29.154)>
   DEBUG  04-14 19:58:09.513  ResponseCandidateStream  Got unexpected BlockBodies payload from ETHPeer <Node(0x053d@79.98.29.154)>
--
   DEBUG  04-14 19:58:09.585   FastChainBodySyncer  Requesting receipts for 195 headers from ETHPeer <Node(0x053d@79.98.29.154)>
   DEBUG  04-14 19:58:10.347   FastChainBodySyncer  Timed out requesting block bodies for 128 headers from ETHPeer <Node(0x053d@79.98.29.154)>
   DEBUG  04-14 19:58:10.351  ResponseCandidateStream  Got unexpected BlockBodies payload from ETHPeer <Node(0x053d@79.98.29.154)>
--
   DEBUG  04-14 19:58:25.358   FastChainBodySyncer  Requesting block bodies for 128 headers from ETHPeer <Node(0x053d@79.98.29.154)>
   DEBUG  04-14 19:58:26.240   FastChainBodySyncer  Timed out requesting block bodies for 128 headers from ETHPeer <Node(0x053d@79.98.29.154)>
   DEBUG  04-14 19:58:26.244  ResponseCandidateStream  Got unexpected BlockBodies payload from ETHPeer <Node(0x053d@79.98.29.154)>
```

**Comment by veox** [comment]
Modifying `trinity/protocol/eth/constants.py` to contain:

``` python
# Max number of items we can ask for in ETH requests. These are the values used
# in geth and if we ask for more than this the peers will disconnect from us.
MAX_STATE_FETCH = 384
MAX_BODIES_FETCH = 64 # half
MAX_RECEIPTS_FETCH = 128 # half
MAX_HEADERS_FETCH = 192
```

... seems to alleviate (but not eradicate) this particular issue. I've had continuous sync for ~105 minutes now.

**Comment by pipermerriam** [comment]
This *might* be justification for implementing dynamic tuning of request sizes based on historical peer data.  I've experienced something that feels similar though not the same as this, consistently seeing responses for receipts come back with significantly fewer than requested.

**Comment by cburgdorf** [comment]
Just noting that I applied the same tuning for my isolated process sync
branch. I thought it was related to my changes but looks like there's more
to it.

**Comment by carver** [comment]
I swear we had this conversation before, but I can't find it. I think a part of the problem is the way this timeout tuning works:
https://github.com/ethereum/trinity/blob/05a1cdee3322d79a049684e2f1d99f7f06912dc6/trinity/protocol/common/managers.py#L100-L109

If the peer response time shifts (say, their connection gets a bit worse) and we have a tight bound on their timeout based on ^ then we will get a bunch of timeouts until the average time increases enough to get a response again. I am not really convinced that this adaptive timeout is buying us anything, and have it patched off in my local, from:
https://github.com/ethereum/trinity/blob/05a1cdee3322d79a049684e2f1d99f7f06912dc6/trinity/protocol/common/managers.py#L97
to
```py
if True: 
```
which improved things for me.

**Comment by pipermerriam** [comment]
Hrm,  I think you're probably right and I'm just emotionally attached to that code.  My inclination is to suggest we just put a lower bound on that value so that the timeout never drops below some reasonable minimum which it currently can.  The only argument I have for this is that it ensures we still timeout quests *promptly* in the event that someone stops responding but I'll admit that argument feels weak.....

**Comment by carver** [comment]
> My inclination is to suggest we just put a lower bound on that value so that the timeout never drops below some reasonable minimum which it currently can.

Eh, it just adds another constant that we don't really know how to choose, and might be wrong depending on your network.

> The only argument I have for this is that it ensures we still timeout quests promptly in the event that someone stops responding but I'll admit that argument feels weak.....

Typically when a peer actually drops off, we get some kind of error right away in the form of a connection lost, so yeah I'm still not really seeing the benefit. Given that it's causing us problems, it seems like reverting to the simpler thing that works is the right first step, and we can add an issue to come back to it later.

**Comment by veox** [comment]
> I swear we had this conversation before, but I can't find it.

@carver It might've been in the `py-evm` repo, or on gitter.

-----

Just had same behaviour on regular full sync (also Ropsten). No detailed log, but here's a grepped snippet:

```
    INFO  04-18 21:19:57.567  RegularChainBodySyncer  Imported block 297441 (0 txs) in 0.02 seconds
(stuck here for an hour as in original report: extremely fast time-out, without recovery; node restarted)
    INFO  04-18 22:36:12.090  RegularChainBodySyncer  Imported block 297442 (63 txs) in 7.18 seconds
    INFO  04-18 22:36:19.462  RegularChainBodySyncer  Imported block 297443 (81 txs) in 7.37 seconds
```

Hypothesis:

1. `trinity` has gone through a streak of blocks light<sup>1</sup> in computation/IO - many 0-tx ones;
2. `trinity` has reduced the allowed time-out to a very low value for that peer;
3. when a "heavy"<sup>1</sup> block comes, the peer times out;
4. ???
5. `trinity` does not adjust time-out.

<sup>1</sup> (by its own measure; there might be yet-unknown inefficiencies that make the scale different than, say, for `geth` or `parity`)

-----

Blind guess:

There may be a bug that counts local-side processing time as part of transmission time, from which the allowed time-out is determined.

Say, processing of "request answered/lapsed, adjust timer if needed" happens after "block imported". But time-out timer terminates block importing, the request is considered unanswered; then timer adjustment finally "happens", but it doesn't differ too much from the target. The request is repeated with the same time-out; the loop goes on forever.

-----

Also:

https://github.com/ethereum/trinity/blob/05a1cdee3322d79a049684e2f1d99f7f06912dc6/trinity/protocol/common/managers.py#L114-L116

It is easily possible that the `timeout` is 0:

``` python
if time.perf_counter() - start_at > outer_timeout:
    timeout_remaining = 0
    payload_timeout = min(inner_timeout, timeout_remaining) # payload_timeout = 0
```

**Comment by veox** [comment]
I suppose this got addressed by PR #568. Therefore closing.
